File: singleton-pattern/go/src/EagerSingleton.go
package src

import "fmt"

// 饿汉模式单例
type EagerSingleton struct {
  name string
}

func (s *EagerSingleton) Run() {
  fmt.Println("EagerSingleton::run()", s.name)
}

// 定义私有变量，用来保存实例
var eagerSingletonInstance *EagerSingleton

// 由主程序在初始化时执行，让实例保存在内存中，这样将获得唯一值
func EagerSingletonInit() {
  eagerSingletonInstance = new(EagerSingleton)
  // eagerSingletonInstance = &EagerSingleton{}
}

func GetEagerSingleton(name string) *EagerSingleton {
  // 如果name为空则赋值，已经赋值过了说明已经实例化了
  if eagerSingletonInstance.name == "" {
    eagerSingletonInstance.name = name
  }
  return eagerSingletonInstance
}

File: singleton-pattern/c/src/eager_singleton.c
#include "func.h"

void eager_singleton_run(EagerSingleton *singleton)
{
  printf("\r\n EagerSingleton::run() [name=%s value=%d]", singleton->name, singleton->value);
}

// 直接用static声明struct对象并初始化，相当于实例化类
static EagerSingleton earge_singleton_instance = {
    .name = "",
    .run = &eager_singleton_run,
    .value = 0,
};

// 饿汉模式，全局唯一，系统加载则初始化变量，无并发安全问题，但可能存在浪费
EagerSingleton *get_eager_singleton_instance(char *name)
{
  // 如果name为空则更新name
  if (strcmp(earge_singleton_instance.name, "") == 0)
  {
    strcpy(earge_singleton_instance.name, name);
  }
  return &earge_singleton_instance;
}